src/comm/TcpAcceptor.cc

```diff
@@ -392,10 +391,18 @@ Comm::TcpAcceptor::oldAccept(Comm::ConnectionPointer &details)
     details->local = *gai;
     Ip::Address::FreeAddr(gai);
 
-    // Perform NAT or TPROXY operations to retrieve the real client/dest IP addresses
-    if (conn->flags&(COMM_TRANSPARENT|COMM_INTERCEPTION) && !Ip::Interceptor.Lookup(details, conn)) {
-        debugs(50, DBG_IMPORTANT, "ERROR: NAT/TPROXY lookup failed to locate original IPs on " << details);
-        return Comm::NOMESSAGE;
+    // Handle interceptedLocally() traffic here. The PROXY protocol supplies original
+    // client and destination IP addresses later, after parsing the PROXY protocol prefix.
+    if (listenPort_->flags.tproxyInterceptLocally()) {
+        if (!Ip::Interceptor.LookupTproxy(details)) {
+             debugs(50, DBG_IMPORTANT, "ERROR: TPROXY lookup failed to locate original IPs on " << details);
+             return Comm::NOMESSAGE;
+         }
```

Comment:
From the API point of view, we have two symmetrical methods here - LookupNat() and LookupTproxy() which, as one would expect, perform lookups, storing results into 'details' (and return false if this lookup fails). The fact that LookupTproxy() does not do this itself (see the comment in Intercept::TproxyTransparent()), trusting the spoofing Squid code (running elsewhere) is a mere implementation detail. In other words, from the caller's perspective, there is nothing criminal in LookupNat() or LookupTproxy() returning false - for otherwise, we should have placed that assertion inside LookupTproxy() itself (and made it return void).

> TPROXY activated.

BTW, this 'activation' should have been checked earlier by Intercept::ProbeForTproxy().

Reply:
> From the API point of view, we have two symmetrical methods here - 

Ah, No. The appearance of symmetry is an artifact of them having been combined behind the Lookup() API, plus Linux 2.x once needed some actions long ago removed. The TPROXY side of things today is actually just a test to prevent NAT lookups changing connection details.

Yes, there are quite a few things we could do to clean this situation up. But IMO that is all out of scope for this PR.

> 
> > TPROXY activated.
> 
> BTW, this 'activation' should have been checked earlier by Intercept::ProbeForTproxy().

Exactly why I suggested the use of assert(). TPROXY either "works" or the port is never opened for listening - in fact right now Squid refuses to startup.

Reply:
The official PR1062 changes have been applied to this PR.